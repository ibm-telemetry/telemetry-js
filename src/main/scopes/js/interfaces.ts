/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
