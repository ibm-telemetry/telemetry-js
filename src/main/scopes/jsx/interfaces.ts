/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type * as ts from 'typescript'

import { type JsxScopeAccumulator } from './jsx-scope-accumulator.js'

export interface Attribute { name: string, value: unknown }

export interface JsxElement {
  name: string
  prefix: string
  raw?: string
  attributes: JsxElementAttribute[]
  pos: number
  end: number
}

export interface JsxElementAttribute {
  name: string
  value: unknown
  isVarReference: boolean
  isSpread: boolean
}

export interface JsxImport {
  importPath: string
  elements: JsxImportElement[]
}

export interface JsxImportElement {
  name: string
  isDefault: boolean
  rename?: string
  isAll: boolean
}

export interface ASTNodeHandler {
  handle: (node: ts.Node, accumulator: JsxScopeAccumulator) => void
}
