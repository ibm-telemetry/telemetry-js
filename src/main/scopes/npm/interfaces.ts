/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export interface PackageData {
  name: string
  version: string
}

export interface InstallingPackage extends PackageData {
  dependencies: PackageData[]
}

export interface DependencyTreeDependency {
  version: string
  dependencies: Record<string, DependencyTreeDependency>
}

export interface DependencyTree extends PackageData {
  name: string
  version: string
  dependencies: Record<string, DependencyTreeDependency>
  [key: string]: unknown
}
