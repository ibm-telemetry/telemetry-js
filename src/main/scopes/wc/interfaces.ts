/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type { ComplexValue } from '../js/complex-value.js'

export interface WcElementAttribute {
  name: string
  value: string | number | boolean | ComplexValue | null | undefined
}

export interface WcElement {
  name: string
  attributes: WcElementAttribute[]
}

/**
 * An interface for generic AST node adapters that provide
 * a unified way to traverse nodes regardless of syntax tree format.
 *
 * @template TNode The type of the original AST node being adapted.
 */
export interface INodeAdapter<TNode> {
  /**
   * Returns the kind/type of the node.
   *
   * @returns The node kind or type identifier.
   */
  getKind(): string | number

  /**
   * Lazily yields child nodes wrapped in the same adapter interface.
   *
   * @returns Iterable of child nodes.
   */
  getChildren(): Iterable<INodeAdapter<TNode>>

  /**
   * Returns the raw original AST node.
   *
   * @returns The unwrapped AST node.
   */
  getNode(): TNode
}
