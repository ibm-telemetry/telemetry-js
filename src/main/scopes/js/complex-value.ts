/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
