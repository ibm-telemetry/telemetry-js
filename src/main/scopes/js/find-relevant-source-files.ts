/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
