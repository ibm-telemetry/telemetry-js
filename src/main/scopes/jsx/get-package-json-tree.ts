/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import fs from 'fs/promises'
import path from 'path'

import { DirectoryEnumerator } from '../../core/directory-enumerator.js'
import { type Logger } from '../../core/log/logger.js'
import { type FileTree } from './interfaces.js'

async function hasPackageJsonFile(dir: string) {
  const packageJsonPath = path.join(dir, 'package.json')

  try {
    await fs.access(packageJsonPath)
  } catch {
    return false
  }

  return true
}

/**
 * Constructs an array of trees from all package.json files in the given repository,
 *  where every nested object is a subpath of the root.
 *
 * @param root - Root-most directory to consider.
 * @param cwd - Current working directory to find tracked files for.
 * @param logger - Logger instance.
 * @returns Array of trees of package.json files.
 */
export async function getPackageJsonTree(
  root: string,
  cwd: string,
  logger: Logger
): Promise<FileTree[]> {
  const dirs = (await new DirectoryEnumerator(logger).find(cwd, root, hasPackageJsonFile)).sort(
    (a, b) => a.split('/').length - b.split('/').length
  )

  const tree: FileTree[] = []

  dirs.forEach((path) => {
    let currNode = tree
    let nextNode = tree.find((root) => path.includes(root.path))
    while (nextNode !== null && nextNode !== undefined) {
      currNode = nextNode.children
      nextNode = currNode.find((node) => path.includes(node.path))
    }
    currNode.push({ path, children: [] })
  })

  return tree
}
