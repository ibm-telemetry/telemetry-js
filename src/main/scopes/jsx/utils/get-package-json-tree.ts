/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'path'

import { type Logger } from '../../../core/log/logger.js'
import { TrackedFileEnumerator } from '../../../core/tracked-file-enumerator.js'
import { type FileTree } from '../interfaces.js'

/**
 * Constructs an array of trees from all package.json files in the given repository,
 * where every nested object is a subpath of the root.
 *
 * @param root - Root-most directory to consider. This is an absolute path.
 * @param logger - Logger instance.
 * @returns Array of trees of package.json files.
 */
export async function getPackageJsonTree(root: string, logger: Logger): Promise<FileTree[]> {
  logger.traceEnter('', 'getPackageJsonTree', [root])

  const dirs = (
    await new TrackedFileEnumerator(logger).find(
      root,
      (fileName) => path.basename(fileName) === 'package.json'
    )
  )
    .map((f) => path.dirname(f))
    .sort((a, b) => a.split('/').length - b.split('/').length)

  const tree: FileTree[] = []

  dirs.forEach((dir) => {
    let currNode = tree
    let nextNode = tree.find((root) => dir.includes(root.path))
    while (nextNode !== null && nextNode !== undefined) {
      currNode = nextNode.children
      nextNode = currNode.find((node) => dir.includes(node.path))
    }
    currNode.push({ path: dir, children: [] })
  })

  logger.traceExit('', 'getPackageJsonTree', tree)
  return tree
}
