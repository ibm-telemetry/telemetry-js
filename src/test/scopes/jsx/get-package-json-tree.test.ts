/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'path'
import { describe, expect, it } from 'vitest'

import { getPackageJsonTree } from '../../../main/scopes/jsx/get-package-json-tree.js'
import { type FileTree } from '../../../main/scopes/jsx/interfaces.js'
import { Fixture } from '../../__utils/fixture.js'
import { initLogger } from '../../__utils/init-logger.js'

describe('getPackageJsonTree', () => {
  const logger = initLogger()
  it('correctly gets a tree', async () => {
    const rootFixture = new Fixture('projects')
    const tree = await getPackageJsonTree(rootFixture.path, logger)

    const setRelativePaths = (tree: FileTree) => {
      tree.path = path.relative(import.meta.url, tree.path)
      tree.children.forEach((child) => {
        setRelativePaths(child)
      })
    }

    tree.forEach((fileTree) => {
      setRelativePaths(fileTree)
    })
    expect(tree).toMatchSnapshot()
  })
  it('empty directory', async () => {
    const rootFixture = new Fixture('jsx-samples')
    const tree = await getPackageJsonTree(rootFixture.path, logger)
    expect(tree).toHaveLength(0)
  })
  it('directory does not exist', async () => {
    const rootFixture = new Fixture('jsx-samples/does-not-exist')
    const tree = await getPackageJsonTree(rootFixture.path, logger)
    expect(tree).toHaveLength(0)
  })
})
