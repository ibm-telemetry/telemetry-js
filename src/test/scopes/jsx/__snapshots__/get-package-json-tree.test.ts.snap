// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPackageJsonTree > correctly gets a tree 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/basic-project/node_modules/foo",
      },
      {
        "children": [],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/basic-project/node_modules/instrumented",
      },
    ],
    "path": "../../../../../../../../../../../src/test/__fixtures/projects/basic-project",
  },
  {
    "children": [
      {
        "children": [],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/package1",
      },
      {
        "children": [
          {
            "children": [],
            "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/package2/node_modules/instrumented",
          },
        ],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/package2",
      },
      {
        "children": [
          {
            "children": [],
            "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/package3/node_modules/instrumented",
          },
        ],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/package3",
      },
      {
        "children": [],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/node_modules/instrumented",
      },
      {
        "children": [],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/node_modules/intermediate",
      },
      {
        "children": [],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy/node_modules/root-only",
      },
    ],
    "path": "../../../../../../../../../../../src/test/__fixtures/projects/complex-nesting-thingy",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "path": "../../../../../../../../../../../src/test/__fixtures/projects/nested-node-modules-folders/node_modules/nested-dep/node_modules/very-nested-dep",
          },
        ],
        "path": "../../../../../../../../../../../src/test/__fixtures/projects/nested-node-modules-folders/node_modules/nested-dep",
      },
    ],
    "path": "../../../../../../../../../../../src/test/__fixtures/projects/nested-node-modules-folders",
  },
  {
    "children": [],
    "path": "../../../../../../../../../../../src/test/__fixtures/projects/no-node-modules-folders",
  },
]
`;
